libs/wingc/src/type_check.rs

```diff
@@ -4485,30 +4493,112 @@ impl<'a> TypeChecker<'a> {
+		let var_type = match &var {
+			ResolveReferenceResult::Variable(var) => var.type_,
+			ResolveReferenceResult::Location(_, elem_type) => *elem_type,
+		};
 
-		if !var.type_.is_unresolved() && !var.reassignable {
-			report_diagnostic(Diagnostic {
-				message: "Variable is not reassignable".to_string(),
-				span: Some(variable.span()),
-				annotations: vec![DiagnosticAnnotation {
-					message: "defined here (try adding \"var\" in front)".to_string(),
-					span: var.name.span(),
-				}],
-				hints: vec![],
-			});
-		} else if var_phase == Phase::Preflight && env.phase == Phase::Inflight {
-			self.spanned_error(variable, "Variable cannot be reassigned from inflight".to_string());
+		// First, check if the assignment violates any mutability restrictions.
+		//
+		// Examples:
+		//
+		// ```
+		// let x = 5;
+		// x = 6; // error: x is not reassignable
+		// ```
+		//
+		// ```
+		// let x = Array<num>[1, 2, 3];
+		// x[0] = 4; // error: cannot update elements of an immutable array
+		// ```
+		match &var {
+			ResolveReferenceResult::Variable(var) => {
+				if !var.reassignable && !var.type_.is_unresolved() {
+					self.spanned_error_with_annotations(
+						variable,
+						"Variable is not reassignable".to_string(),
+						vec![DiagnosticAnnotation {
+							message: "defined here (try adding \"var\" in front)".to_string(),
+							span: var.name.span(),
+						}],
+					);
+				} else if var_phase == Phase::Preflight && env.phase == Phase::Inflight {
+					self.spanned_error(variable, "Variable cannot be reassigned from inflight".to_string());
+				}
+			}
+			ResolveReferenceResult::Location(container_type, _) => match **container_type {
+				Type::Anything | Type::MutJson | Type::MutArray(_) | Type::MutMap(_) => {}
+				Type::Map(_) => {
+					self.spanned_error_with_hints(
+						variable,
+						"Cannot update elements of an immutable Map".to_string(),
+						vec!["Consider using MutMap instead".to_string()],
+					);
+				}
+				Type::Json(_) => {
+					self.spanned_error_with_hints(
+						variable,
+						"Cannot update elements of an immutable Json".to_string(),
+						vec!["Consider using MutJson instead".to_string()],
+					);
+				}
+				Type::Array(_) => {
+					self.spanned_error_with_hints(
+						variable,
+						"Cannot update elements of an immutable Array".to_string(),
+						vec!["Consider using MutArray instead".to_string()],
+					);
+				}
+				Type::String => {
+					self.spanned_error(variable, "Strings are immutable".to_string());
+				}
+				Type::Inferred(_)
+				| Type::Unresolved
+				| Type::Number
+				| Type::Duration
+				| Type::Boolean
+				| Type::Void
+				| Type::Nil
+				| Type::Optional(_)
+				| Type::Set(_)
+				| Type::MutSet(_)
+				| Type::Function(_)
+				| Type::Class(_)
+				| Type::Interface(_)
+				| Type::Struct(_)
+				| Type::Enum(_) => self.spanned_error(
+					variable,
+					format!("Unsupported reassignment of element of type {}", var_type),
```

Comment:
Was there a test for this?

Reply:
I didn't, but I can try to add one. It's actually kinda tricky coming up with examples that produce this error, but it seems to correspond to degenerate cases like:

```
let var x = 3 + "hello";
x["hi"] = 4; // error: Unsupported reassignment of element of type unresolved
```